src/main/java/net/minecraftforge/common/util/MutableHashedLinkedList.java

```diff
@@ -0,0 +1,205 @@
+/*
+ * Copyright (c) Forge Development LLC and contributors
+ * SPDX-License-Identifier: LGPL-2.1-only
+ */
+package net.minecraftforge.common.util;
+
+import java.util.Iterator;
+import java.util.Map;
+import java.util.NoSuchElementException;
+
+import org.apache.commons.lang3.mutable.MutableObject;
+import org.jetbrains.annotations.Nullable;
+
+import it.unimi.dsi.fastutil.Hash.Strategy;
+import it.unimi.dsi.fastutil.objects.Object2ObjectOpenCustomHashMap;
+
+public class MutableHashedLinkedList<K, V> implements Iterable<Map.Entry<K, V>>
+{
+    private final Map<K, Entry> entries;
+    private final MergeFunction<K, V> merge;
+    private final MutableObject<Entry> head = new MutableObject<>();
+    private final MutableObject<Entry> last = new MutableObject<>();
+
+    public MutableHashedLinkedList(Strategy<? super K> strategy) {
+        this(strategy, (k, v1, v2) -> v2);
+    }
+
+    public MutableHashedLinkedList(Strategy<? super K> strategy, MergeFunction<K, V> merge) {
+        this.entries = new Object2ObjectOpenCustomHashMap<>(strategy);
+        this.merge = merge;
+    }
+
+    public V put(K key, V value) {
+        var old = entries.get(key);
+        if (old != null) {
+            V ret = old.value;
+            old.value = merge.apply(key, ret, value);
+            return ret;
+        }
+
+        var self = new Entry(key, value);
+
+        var l = last.getValue();
+        self.previous = l;
+        if (l == null)
+            head.setValue(self);
+        else
+            l.next = self;
+        last.setValue(self);
+
+        entries.put(key, self);
+        return null;
+    }
+
+    public boolean contains(K key) { return this.entries.containsKey(key); }
+    public boolean isEmpty() { return this.entries.isEmpty(); }
+
+    @Nullable
+    public V remove(K key) {
+        var ret = this.entries.remove(key);
+        if (ret == null)
+            return null;
+
+        ret.remove();
+        return ret.getValue();
+    }
+
+    @Nullable
+    public V get(K key) {
+        var entry = entries.get(key);
+        return entry == null ? null : entry.getValue();
+    }
+
+    @Override
+    public Iterator<Map.Entry<K, V>> iterator() {
+        return new Iterator<Map.Entry<K, V>>() {
```

Comment:
This can use the diamond operator instead.
```suggestion
        return new Iterator<>() {
```